--- busybox-1.36.1.orig/coreutils/stat.c	2025-10-07 15:03:01.000000000 +0000
+++ busybox-1.36.1/coreutils/stat.c	2023-05-18 22:34:17.000000000 +0000
@@ -106,6 +106,9 @@
 
 #include "libbb.h"
 #include "common_bufsiz.h"
+#ifdef __QNX__
+# include <sys/mount.h>
+#endif
 
 enum {
 	OPT_TERSE       = (1 << 0),
@@ -227,7 +230,7 @@
 
 /* "man statfs" says that statfsbuf->f_fsid is a mess */
 /* coreutils treats it as an array of ints, most significant first */
-static unsigned long long get_f_fsid(const struct statfs *statfsbuf)
+static unsigned long long get_f_fsid(const struct statvfs *statfsbuf)
 {
 	const unsigned *p = (const void*) &statfsbuf->f_fsid;
 	unsigned sz = sizeof(statfsbuf->f_fsid) / sizeof(unsigned);
@@ -260,7 +263,7 @@
 		const char *const filename, const void *data
 		IF_SELINUX(, security_context_t scontext))
 {
-	const struct statfs *statfsbuf = data;
+	const struct statvfs *statfsbuf = data;
 	if (m == 'n') {
 		printfs(pformat, filename);
 	} else if (m == 'i') {
@@ -268,12 +271,12 @@
 		printf(pformat, get_f_fsid(statfsbuf));
 	} else if (m == 'l') {
 		strcat(pformat, "lu");
-		printf(pformat, (unsigned long) statfsbuf->f_namelen);
+		printf(pformat, (unsigned long) statfsbuf->f_namemax);
 	} else if (m == 't') {
 		strcat(pformat, "lx");
-		printf(pformat, (unsigned long) statfsbuf->f_type); /* no equiv */
+		printf(pformat, statfsbuf->f_basetype); /* no equiv */
 	} else if (m == 'T') {
-		printfs(pformat, human_fstype(statfsbuf->f_type));
+		printfs(pformat, statfsbuf->f_basetype);
 	} else if (m == 'b') {
 		strcat(pformat, "llu");
 		printf(pformat, (unsigned long long) statfsbuf->f_blocks);
@@ -470,7 +473,7 @@
 #endif
 static bool do_statfs(const char *filename, const char *format)
 {
-	struct statfs statfsbuf;
+	struct statvfs statfsbuf;
 #if !ENABLE_FEATURE_STAT_FORMAT
 	const char *format;
 #endif
@@ -488,7 +491,7 @@
 		}
 	}
 #endif
-	if (statfs(filename, &statfsbuf) != 0) {
+	if (statvfs(filename, &statfsbuf) != 0) {
 		bb_perror_msg("can't read file system information for '%s'", filename);
 		return 0;
 	}
--- busybox-1.36.1.orig/include/libbb.h	2025-10-07 15:03:01.000000000 +0000
+++ busybox-1.36.1/include/libbb.h	2023-05-18 22:34:17.000000000 +0000
@@ -1834,7 +1834,7 @@
 char *is_in_ino_dev_hashtable(const struct stat *statbuf) FAST_FUNC;
 void add_to_ino_dev_hashtable(const struct stat *statbuf, const char *name) FAST_FUNC;
 void reset_ino_dev_hashtable(void) FAST_FUNC;
-#ifdef __GLIBC__
+#if defined(__GLIBC__) || defined(__QNX__)
 /* At least glibc has horrendously large inline for this, so wrap it */
 unsigned long long bb_makedev(unsigned major, unsigned minor) FAST_FUNC;
 #undef makedev
--- busybox-1.36.1.orig/include/platform.h	2025-10-07 15:03:01.000000000 +0000
+++ busybox-1.36.1/include/platform.h	2023-05-18 22:34:17.000000000 +0000
@@ -164,6 +164,8 @@
 # define bswap_64 __bswap64
 # define bswap_32 __bswap32
 # define bswap_16 __bswap16
+# elif defined(__QNX__)
+#  include <sys/param.h>
 #else
 # include <byteswap.h>
 # include <endian.h>
@@ -490,6 +492,21 @@
 # undef HAVE_PRINTF_PERCENTM
 #endif
 
+#if defined(__QNX__)
+# undef HAVE_MNTENT_H
+# undef HAVE_SYS_STATFS_H
+# undef HAVE_UNLOCKED_LINE_OPS
+# undef HAVE_STRVERSCMP
+# undef HAVE_STRCHRNUL
+# undef HAVE_DPRINTF
+# undef HAVE_MEMPCPY
+# undef HAVE_MEMRCHR
+# define SA_RESTART 0
+# ifndef __X86__
+#  undef HAVE_GETLINE
+# endif
+#endif
+
 #if defined(__dietlibc__)
 # undef HAVE_STRCHRNUL
 #endif
--- busybox-1.36.1.orig/libbb/makedev.c	2025-10-07 15:03:01.000000000 +0000
+++ busybox-1.36.1/libbb/makedev.c	2023-05-18 22:34:17.000000000 +0000
@@ -11,7 +11,7 @@
 
 /* Different Unixes want different headers for makedev */
 #if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) \
- || defined(__APPLE__)
+ || defined(__APPLE__) || defined(__QNX__)
 # include <sys/types.h>
 #else
 # include <features.h>
@@ -28,4 +28,11 @@
 {
 	return makedev(major, minor);
 }
+#elifdef __QNX__
+/* Suppress gcc "no previous prototype" warning */
+unsigned long long FAST_FUNC bb_makedev(unsigned major, unsigned minor);
+unsigned long long FAST_FUNC bb_makedev(unsigned major, unsigned minor)
+{
+       return makedev(0, major, minor);
+}
 #endif
--- busybox-1.36.1.orig/scripts/kconfig/mconf.c	2025-10-07 15:03:01.000000000 +0000
+++ busybox-1.36.1/scripts/kconfig/mconf.c	2023-05-18 22:34:17.000000000 +0000
@@ -475,7 +475,7 @@
 		struct sigaction sa;
 		sa.sa_handler = winch_handler;
 		sigemptyset(&sa.sa_mask);
-		sa.sa_flags = SA_RESTART;
+		sa.sa_flags = 0;
 		sigaction(SIGWINCH, &sa, NULL);
 	}
 #endif
